python/lsst/daf/butler/queries/visitors.py

```diff
@@ -0,0 +1,540 @@
+            Expression to visit.
+
+        Returns
+        -------
+        result : `object`
+            Implementation-defined.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def visit_dimension_key_reference(self, expression: tree.DimensionKeyReference) -> _T:
+        """Visit a column expression that represents a dimension column.
+
+        Parameters
+        ----------
+        expression : `tree.DimensionKeyReference`
+            Expression to visit.
+
+        Returns
+        -------
+        result : `object`
+            Implementation-defined.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def visit_dimension_field_reference(self, expression: tree.DimensionFieldReference) -> _T:
+        """Visit a column expression that represents a dimension record field.
+
+        Parameters
+        ----------
+        expression : `tree.DimensionFieldReference`
+            Expression to visit.
+
+        Returns
+        -------
+        result : `object`
+            Implementation-defined.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def visit_dataset_field_reference(self, expression: tree.DatasetFieldReference) -> _T:
+        """Visit a column expression that represents a dataset field.
+
+        Parameters
+        ----------
+        expression : `tree.DatasetFieldReference`
+            Expression to visit.
+
+        Returns
+        -------
+        result : `object`
+            Implementation-defined.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def visit_unary_expression(self, expression: tree.UnaryExpression) -> _T:
+        """Visit a column expression that represents a unary operation.
+
+        Parameters
+        ----------
+        expression : `tree.UnaryExpression`
+            Expression to visit.
+
+        Returns
+        -------
+        result : `object`
+            Implementation-defined.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def visit_binary_expression(self, expression: tree.BinaryExpression) -> _T:
+        """Visit a column expression that wraps a binary operation.
+
+        Parameters
+        ----------
+        expression : `tree.BinaryExpression`
+            Expression to visit.
+
+        Returns
+        -------
+        result : `object`
+            Implementation-defined.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def visit_reversed(self, expression: tree.Reversed) -> _T:
+        """Visit a column expression that switches sort order from ascending
+        to descending.
+
+        Parameters
+        ----------
+        expression : `tree.Reversed`
+            Expression to visit.
+
+        Returns
+        -------
+        result : `object`
+            Implementation-defined.
+        """
+        raise NotImplementedError()
+
+
+class PredicateVisitor(Generic[_A, _O, _L]):
```

Comment:
One of the more important visitors I anticipate is one whose return values are `sqlalchemy.ColumnElement[bool]`, and the more concrete one can't do that.  There are three generic arguments instead of one because I recalled needing that on a similar visitor class in the registry query system, but I don't recall why.  I'd like to leave it until we have a concrete driver or two.